### PR TITLE
Parameter scheduling handler

### PR DESCRIPTION
**Pull request by sean-adler**
https://github.com/pytorch/ignite/issues/169

Adds some handlers for optimizer parameter scheduling.

The proposed API looks like this:

```
model = Model()
optimizer = SGD(model.parameters(), lr=1e-3)
trainer = create_supervised_trainer(model, optimizer, F.nll_loss)

scheduler = LinearScheduler(optimizer, 'lr', 1e-3, 1e-1, len(train_loader))
trainer.add_event_handler(Events.ITERATION_COMPLETED, scheduler)

# Linearly increases LR from 1e-3 to 1e-1 and back to 1e-3 over the course of 1 epoch
```

* the `ParamScheduler` base class doesn't do much other than setting up the basic API and (optionally) saving parameter histories.
* the `CyclicalScheduler` class implements `get_cycle_progress()`, which returns a percentage representing how far we are through the current cycle. This makes it fairly easy to implement cyclical parameter schedules - for example, the cosine annealing algorithm as described in the [SGDR](https://arxiv.org/pdf/1608.03983.pdf) paper becomes 2 lines of code.
* the `CosineAnnealingScheduler` class implements cosine annealing as described in that paper, and the `LinearScheduler` class implements the "triangular" linear cyclical learning rate policy as described in [Cyclical Learning Rates for Training Neural Networks](https://arxiv.org/pdf/1506.01186.pdf).

Random side-note: I don't love the amount of OOP here (3 layers of inheritance is kind of a lot), but I wanted the user-facing API of writing new schedules (e.g. `LinearScheduler.get_param`) to be simple, so hopefully the rest of the `param_scheduler` module isn't too convoluted.

### CODE REVIEW

**Comment by vfdev-5** [comment]
As it was discussed in #169 @sean-adler could you please put this code and tests into `contrib` module ? 
If you do not mind I'll close soon this PR in favor of a new one to `contrib`

**Comment by vfdev-5** [comment]
@sean-adler I updated the code if you do not mind

**Comment by vfdev-5** [comment]
@alykhantejani @jasonkriss review please

**Comment by jasonkriss** [comment]
@sean-adler Nice work! I think this looks really clean. Nice work on the tests! The only thing I would request is that we add a docstring for `get_param`. I also agree that the multiple layers of inheritance is less than ideal. If anyone has a better alternative, that'd be great. Otherwise, I think it's fine for now.

**Comment by vfdev-5** [comment]
@jasonkriss I added `get_param` docstring and renamed `LinearScheduler` to `LinearCyclicalScheduler`

**Comment by jasonkriss** [comment]
👍thanks @vfdev-5. LGTM.

**Comment by vfdev-5** [comment]
@jasonkriss could we merge this, please ?
